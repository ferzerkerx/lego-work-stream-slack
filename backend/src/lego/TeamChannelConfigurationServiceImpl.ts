import {
  AppEvent,
  AppEventTypes,
  ApplicationEventEmitter,
  TeamChannelConfigurationRepository,
  TeamChannelConfigurationService,
} from './Types';
import { TeamChannelConfiguration } from './LegoMessageFactory';
import { SlackBot } from 'botkit';
import { ErrorUtil } from '../utils/ErrorUtil';

export class TeamChannelConfigurationServiceImpl
  implements TeamChannelConfigurationService {
  constructor(
    private teamChannelConfigurationRepository: TeamChannelConfigurationRepository,
    private applicationEventEmitter: ApplicationEventEmitter
  ) {}

  update(
    configStr: string,
    bot: SlackBot,
    channelId: string
  ): Promise<TeamChannelConfiguration> {
    const configuration = TeamChannelConfigurationServiceImpl.parseConfig(
      configStr
    );

    if (!configuration) {
      return Promise.reject('Invalid configuration');
    }

    configuration.id = channelId;
    configuration.channelName = channelId;

    return this.teamChannelConfigurationRepository
      .save(configuration)
      .then(configuration => {
        const updatedConfigEvent: AppEvent = {
          name: AppEventTypes.TEAM_CHANNEL_CONFIG_UPDATED,
          data: { configuration: configuration, bot: bot },
        };
        this.applicationEventEmitter.emitEvent(updatedConfigEvent);
        return configuration;
      });
  }

  static parseConfig(configStr: string): TeamChannelConfiguration {
    try {
      const configuration: TeamChannelConfiguration = JSON.parse(configStr);
      if (this.isValid(configuration)) {
        return configuration;
      }
    } catch (e) {
      ErrorUtil.defaultErrorHandling(e);
    }
    return null;
  }

  static isEmpty(str: string | Array<any>) {
    return !str || str.length === 0;
  }

  static isValid(configuration: TeamChannelConfiguration): boolean {
    if (this.isEmpty(configuration.actionDescriptors)) {
      return false;
    }

    for (let descriptor of configuration.actionDescriptors) {
      if (this.isEmpty(descriptor.name) || this.isEmpty(descriptor.text)) {
        return false;
      }
    }

    if (!isFinite(configuration.max) || configuration.max < 0) {
      return false;
    }

    if (!isFinite(configuration.min) || configuration.min < 0) {
      return false;
    }

    if (configuration.max < configuration.min) {
      return false;
    }

    return true;
  }
}
