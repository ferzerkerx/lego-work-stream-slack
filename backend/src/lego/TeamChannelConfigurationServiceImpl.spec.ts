import { TeamChannelConfigurationServiceImpl } from './TeamChannelConfigurationServiceImpl';

function expectInvalidConfiguration(jsonConfig) {
  const configStr: string = JSON.stringify(jsonConfig);

  const configuration = TeamChannelConfigurationServiceImpl.parseConfig(
    configStr,
  );
  expect(configuration).toBeNull();
}

describe('TeamChannelConfigurationServiceImpl', () => {
  test('parseConfig Invalid json', () => {
    const jsonConfig: string = 'invalidJSON';

    expectInvalidConfiguration(jsonConfig);
  });

  test('parseConfig invalid actionDescriptors', () => {
    const jsonConfig = {
      actionDescriptors: [{ name: '', text: '' }],
      min: 0,
      max: 8,
    };

    expectInvalidConfiguration(jsonConfig);
  });

  test('parseConfig empty actionDescriptors', () => {
    const jsonConfig = {
      actionDescriptors: [],
      min: 0,
      max: 8,
    };

    expectInvalidConfiguration(jsonConfig);
  });

  test('parseConfig invalid max', () => {
    const jsonConfig = {
      actionDescriptors: [],
      min: 0,
      max: -19.6,
    };

    expectInvalidConfiguration(jsonConfig);
  });

  test('parseConfig invalid min', () => {
    const jsonConfig = {
      actionDescriptors: [],
      min: -19.6,
      max: 19,
    };

    expectInvalidConfiguration(jsonConfig);
  });

  test('parseConfig invalid range', () => {
    const jsonConfig = {
      actionDescriptors: [],
      min: 20,
      max: 19,
    };

    expectInvalidConfiguration(jsonConfig);
  });

  test('parseConfig valid', () => {
    const jsonConfig = {
      actionDescriptors: [
        { name: 'green', text: 'Green' },
        { name: 'red', text: 'Red' },
        { name: 'orange', text: 'Orange' },
        { name: 'black', text: 'Black' },
      ],
      min: 0,
      max: 8,
    };

    const configStr: string = JSON.stringify(jsonConfig);

    const configuration = TeamChannelConfigurationServiceImpl.parseConfig(
      configStr
    );
    expect(configuration).toMatchSnapshot();
  });
});
