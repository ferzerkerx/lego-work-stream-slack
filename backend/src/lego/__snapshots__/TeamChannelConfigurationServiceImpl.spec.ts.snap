// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamChannelConfigurationServiceImpl parseConfig valid 1`] = `
Object {
  "actionDescriptors": Array [
    Object {
      "name": "green",
      "text": "Green",
    },
    Object {
      "name": "red",
      "text": "Red",
    },
    Object {
      "name": "orange",
      "text": "Orange",
    },
    Object {
      "name": "black",
      "text": "Black",
    },
  ],
  "max": 8,
  "min": 0,
}
`;
